### ISSUE TITLE
Support DSMR 4.2

### ISSUE THREAD

**Issue by Vaoske** [ebaauw/homebridge-p1#17]
Hello,

I can't add any accessories within my Eve app, homebridge normally starts but I got this error:
[2/28/2019, 4:15:54 PM] [P1] Initializing P1 platform...
[2/28/2019, 4:15:54 PM] [P1] homebridge-p1 v0.2.7, node v10.15.1, homebridge v0.4.46
[2/28/2019, 4:15:54 PM] Loading 0 accessories...

Greets from the Netherlands
Vaoske

**Comment by ebaauw** [comment]
I don’t see any error.  The “Loading 0 accessories” meesage is homebridge’s way of telling you haven’t configured any accessory plugins in config.json.

**Comment by Vaoske** [comment]
Thanks for the reply!
The problem is when adding homebridge to iOS only the gateway apear.

In dutch:
Dank u voor uw antwoord. Probleem is wanneer mijn homebridge gateway wordt toegevoegd in woning, geen accessoires zichtbaar zijn. Ofwel ik kan niets aflezen.

Gr
Vaoske

**Comment by ebaauw** [comment]
Could you please create a debug log file and attach it, see https://github.com/ebaauw/homebridge-hue#debug-log-file?  What does `serialport-list -f json` give?

**Comment by Vaoske** [comment]
Hi,

debug log
[homebridge.log](https://github.com/ebaauw/homebridge-p1/files/2916671/homebridge.log)

serialport-list -f joon
[{"comName":"/dev/ttyS1"},{"comName":"/dev/ttyS2"},{"comName":"/dev/ttyS3"},{"comName":"/dev/ttyS4"},{"comName":"/dev/ttyS5"},{"comName":"/dev/ttyS6"},{"comName":"/dev/ttyS7"},{"manufacturer":"FTDI","serialNumber":"AC2EZ4E0","pnpId":"usb-FTDI_FT232R_USB_UART_AC2EZ4E0-if00-port0","vendorId":"0403","productId":"6001","comName":"/dev/ttyUSB0"},{"comName":"/dev/ttyS0"}]

**Comment by ebaauw** [comment]
Thanks.

```
[2/28/2019, 8:50:10 PM] [P1] telegram:
/KFM5KAIFA-METER

1-3:0.2.8(42)
0-0:1.0.0(190228205019W)
0-0:96.1.1(4530303235303030303737373730303136)
1-0:1.8.1(005945.778*kWh)
1-0:1.8.2(004415.532*kWh)
1-0:2.8.1(000212.801*kWh)
1-0:2.8.2(000636.089*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(02.992*kW)
1-0:2.7.0(00.000*kW)
0-0:96.7.21(00001)
0-0:96.7.9(00001)
1-0:99.97.0(2)(0-0:96.7.19)(180824143702S)(0000005600*s)(000101000001W)(2147483647*s)
1-0:32.32.0(00000)
1-0:32.36.0(00000)
0-0:96.13.1()
0-0:96.13.0()
1-0:31.7.0(012*A)
1-0:21.7.0(02.992*kW)
1-0:22.7.0(00.000*kW)
0-2:24.1.0(003)
0-2:96.1.0(4730303139333430333231323232383136)
0-2:24.2.1(190228200000W)(03514.244*m3)
!4983

[2/28/2019, 8:50:10 PM] [P1] raw data: {"type":"KFM5KAIFA-METER","checksum":"4983","version":"4.2","lastupdated":"2019-02-28T20:50:19","id":"E0025000077770016","consumption_low":5945.778,"consumption_normal":4415.532,"consumption_back_low":212.801,"consumption_back_normal":636.089,"tariff":"normal","power":2992,"power_back":0,"failures_short":1,"failures_long":1,"log":{"2018-08-24T14:37:02":5600,"2000-01-01T00:00:01":2147483647},"l1_sags":0,"l1_swells":0,"msg_num":null,"msg_text":null,"l1_current":12,"l1_power":2992,"l1_power_back":0,"d2_type":"gas","d2_id":"G0019340321222816","d2_lastupdated":"2019-02-28T20:00:00","d2_consumption":3514.244}
[2/28/2019, 8:50:10 PM] [P1] data: {"type":"KFM5KAIFA-METER","version":"4.2","msg_text":null,"msg_num":null,"electricity":{"id":"E0025000077770016","lastupdated":"2019-02-28T20:50:19","tariff":"normal","consumption":{"low":5945.778,"normal":4415.532},"power":2992,"failures":{"short":1,"long":1,"log":{"2018-08-24T14:37:02":5600,"2000-01-01T00:00:01":2147483647}},"l1":{"sags":0,"swells":0,"current":12,"power":2992}},"electricityBack":{"id":"E0025000077770016B","lastupdated":"2019-02-28T20:50:19","tariff":"normal","consumption":{"low":212.801,"normal":636.089},"power":0,"l1":{"power":0}},"undefined":{"id":"G0019340321222816","lastupdated":"2019-02-28T20:00:00","consumption":3514.244}}
[2/28/2019, 8:50:10 PM] [P1] KFM5KAIFA-METER v4.2
```
The good news: homebridge-p1 is getting the data from the smart meter.  This is the first time I encounter a DSMR v4.2 meter.  Also, the first time for (I assume) a solar panel installation.  I've been wondering how to expose that, see #3.

```
[2/28/2019, 8:49:57 PM] [P1] listening on /dev/ttyUSB0
[2/28/2019, 8:50:02 PM] [P1] /dev/ttyUSB0: no valid data received
```
The bad news: homebridge-p1 is getting the data too late (after the homebridge server has started), so it doesn't expose the accessories.
```
[2/28/2019, 8:50:20 PM] [P1] Electricity: set homekit total consumption from 10361.31 kWh to 10361.319 kWh
[2/28/2019, 8:50:20 PM] [P1] Electricity: set homekit current consumption from 2992 W to 2948 W
[2/28/2019, 8:50:30 PM] [P1] Electricity: set homekit total consumption from 10361.319 kWh to 10361.327 kWh
[2/28/2019, 8:50:30 PM] [P1] Electricity: set homekit current consumption from 2948 W to 2981 W
```
It looks like DSMR v4.2 is sending a telegram only every 10 seconds; so the default timeout of 5 seconds wouldn't suffice.  Could you set `"timeout": 50` in config.json (that's the value I use for DSMR v2.2) and capture another log file?

```
[2/28/2019, 8:50:10 PM] [P1] Error: This callback function has already been called by someone else; it can only be called one time.
```
Oops.  Looks like that "someone else" was the timeout.

**Comment by ebaauw** [comment]
Found a bug that causes your gas meter not to be exposed.

**Comment by ebaauw** [comment]
> This is the first time I encounter a DSMR v4.2 meter.

No, it's not, see https://github.com/ebaauw/homebridge-p1/issues/7#issuecomment-387366573.  This is a different brand of meter, though.  Maybe the frequency of the telegrams differs per brand, or maybe the other meter is still running an old version of homebridge-p1, from before the timeout was implemented.

**Comment by Vaoske** [comment]
I have an solar panel installation. With "timeout": 50 it works for the electricity!
Updated p1 but no gas meter exposed.

[homebridge.log](https://github.com/ebaauw/homebridge-p1/files/2918760/homebridge.log)

**Comment by ebaauw** [comment]
> Updated p1 but no gas meter exposed.

Just released v0.2.8.

> I have an solar panel installation.

Would appreciate if you share your thoughts on how best to expose this in #3.

**Comment by Vaoske** [comment]
v0.2.8 works!
